Getting started tutorial part two (#8317)
backstage/backstage: docs/getting-started/configuration.md

---

> id: index

Reviewer: ```suggestion
id: configuration
```

Reviewer: Needs a corresponding entry in `microsite/sidebars.json` and `mkdocs.yml` for the page to show up on the sidebar. :)

title: Getting Started, configuring Backstage
description: Getting started with your initial Backstage configuration
---

This is part two of the Getting Started documentation of Backstage. The steps in
this tutorial assume you've installed Backstage app from the npm repository,
like in the [Getting Started guide](./index.md) and want to configure Backstage.

At the end of this tutorial, you can expect:

- Backstage to use a PostgreSQL database
- You'll authenticate using one of the auth providers
- The Backstage GitHub integration to be configured
- You're able to use Software Templates

### Prerequisites

- Access to a Linux-based operating system, such as Linux, MacOS or
  [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/)

> - An account with elevated rights to install prerequisites

Reviewer: I got confused with "account" a bit as it could be GitHub account, Backstage account. But I think this means an `OS user`, doesn't it?

- If the database is not hosted on the same server as the Backstage app, the
  PostgreSQL port needs to be accessible (the default is 5432 or 5433)

> ### Install and configure PostgreSQL

Reviewer: For Mac users, https://postgresapp.com is a single click installation and works automagically. Probably worth adding here along with the apt-get based installation :)

Author: Yep, It's listed in the installation options for Mac on the official site.

Stop Backstage, and go to the root directory of your freshly installed Backstage
App. Use the following commands to start the PostgreSQL client installation:

```shell
# From your Backstage root directory
cd packages/backend
yarn add pg
```

Use your favorite editor to open `app-config.yaml` and add your PostgreSQL
configuration. in the root directory of your Backstage app using the credentials
from the previous steps.

```diff
backend:
  database:
-    client: sqlite3
-    connection: ':memory:'
+    # config options: https://node-postgres.com/api/client
+    client: pg
+    connection:
+      host: ${POSTGRES_HOST}
+      port: ${POSTGRES_PORT}
+      user: ${POSTGRES_USER}
+      password: ${POSTGRES_PASSWORD}
+      # https://node-postgres.com/features/ssl
+      #ssl: require # see https://www.postgresql.org/docs/current/libpq-ssl.html Table 33.1. SSL Mode Descriptions (e.g. require)
+        #ca: # if you have a CA file and want to verify it you can uncomment this section
+        #$file: <file-path>/ca/server.crt
```

You'll use the connection details from the previous step. You can either set the
`POSTGRES_` environment variables prior to launching Backstage, or remove the

> `${...}` values and set actual values in this configuration file.

Reviewer: TIP: Do not add passwords to `app-config.yaml` as the file can be accidently committed. Use `app-config.local.yaml` or other gitignored files for secrets. (Equivalent to `.env` files).

Author: That's true, the getting-started shouldn't be the base for going anywhere official though. 
How to configure B. "for production" Is a much-asked-for-doc. Love to see this pop up there, or separate doc to link too.

The default port for PostgreSQL is `5432` or `5433`, and the host name could be
`127.0.0.1` if installed locally. A word of caution: In general, using
connection details in a configuration file is not recommended.

Start the backstage app:

```shell
yarn dev
```

After Backstage is completely started you'll notice the catalog is populated
with the information, still coming from the configuration files. If you add a
new component, or register an existing one it will be saved in the database.
Later in this tutorial you'll add a service, and you can test if it's persistent
as advertised.

If you want to read more about the database configuration, here's some helpful
links:

- [Configuring Plugin Databases](https://backstage.io/docs/tutorials/configuring-plugin-databases#privileges)
- [Read more about Knex](http://knexjs.org/), which is the library we use for
  the database backend

### Setting up authentication

There's multiple authentication providers available for you to use with
Backstage, feel free to follow

> [the instructions for adding authentication](https://backstage.io/docs/auth/).

Author: Oh my, yes! Thank you for spotting this!

For this tutorial we choose to use GitHub, a free service most of you might be
familiar with. For other options, see
[the auth provider documentation](../auth/github/provider#create-an-oauth-app-on-github).

Go to
[https://github.com/settings/applications/new](https://github.com/settings/applications/new)
to create your OAuth App. The `Homepage URL` should point to Backstage's
frontend, in our tutorial it would be `http://127.0.0.1:3000`. The
`Authorization callback URL` will point to the auth backend, which will most
likely be `http://127.0.0.1:7007/api/auth/github/handler/frame`.

<p align='center'>

>   <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page' />
```

Docusarus 2 is complaining. Ref https://github.com/backstage/backstage/pull/8356

</p>

> Take note of the `Client ID` and the `Client Secret`. Open 'app-config.yaml',

Reviewer: ```suggestion
Take note of the `Client ID` and the `Client Secret`. Open `app-config.yaml`,
```

and add your `clientId` and `clientSecret` to this file. It should end up
looking like this:

```

> auth:
>   # see https://backstage.io/docs/auth/ to learn about auth providers
>   providers:
>     github:
>       development:

Reviewer: ```suggestion
auth:
  # see https://backstage.io/docs/auth/ to learn about auth providers
  environment: development
  providers:
    github:
      development:
```

Having the environment will justify why `github:` is nested as well. :D

Author: :')!

To learn more about Authentication in Backstage, there's the following docs you
could read:

- [Adding Authentication](../auth/)
- [Adding a new Authentication Provider](../auth/add-auth-provider)
- [Using authentication and identity](../auth/using-auth)
- [Using organizational data from GitHub](../integrations/github/org.md)

### Setting up a GitHub Integration

The GitHub integration supports loading catalog entities from GitHub or GitHub
Enterprise. Entities can be added to static catalog configuration, registered
with the catalog-import plugin, or discovered from a GitHub organization. Users
and Groups can also be loaded from an organization. While using GitHub Apps
might be the best way to set up integrations, for this tutorial you'll use a
Personal Access Token.

Create your Personal Access Token by opening the
[the GitHub token creation page](https://github.com/settings/tokens/new). Use a
name to identify this token and put it in the notes field. Choose a number of
days for expiration. If you have a hard time picking a number, we suggest to go
for 7 days, it's a lucky number.

<p align='center'>

>   <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page' />
```

</p>

Set the scope to your likings. For this tutorial, selecting "repo" should be
enough.

In the `app-config.yaml`, search for `integrations:` and add your token, like we
did in below example:

```
integrations:
  github:
    - host: github.com

>       token: ghp_urtokendeinfewinfiwebfweb

Reviewer: ![](https://media.giphy.com/media/3knKct3fGqxhK/giphy.gif)

Author: H4xxorz in ur config

```

That's settled. This information will be leveraged by other plugins.

Some helpful links, for if you want to learn more about:

- [Other available integrations](../integrations/)
- [Using GitHub Apps instead of a Personal Access Token](../plugins/github-apps#docsNav)

### Explore what we've done so far

Open your Backstage frontend. You should see your login screen if you're not
logged in yet. As soon as you've logged in, go to Settings, you'll see your

> profile. Hopefully You'll recognize the profile picture and name on your screen,

Reviewer: ```suggestion
profile. Hopefully you'll recognize the profile picture and name on your screen,
```

otherwise something went terribly wrong.

> Register an existing component

Reviewer: ```suggestion
## Register an existing component
```

??

- Register a new component, by going to `create` and choose
  `Register existing component`

<p align='center'>

>   <img src='../assets/getting-started/b-existing-1.png' alt='Software template main screen, with a blue button to add an existing component'>

Reviewer: Protip, if you want images to be zoomable on click. Example https://backstage.io/docs/features/techdocs/architecture 😁 

```suggestion
  <img data-zoomable src='../assets/getting-started/b-existing-1.png' alt='Software template main screen, with a blue button to add an existing component'>
```

Reviewer: ```suggestion
  <img src='../assets/getting-started/b-existing-1.png' alt='Software template main screen, with a blue button to add an existing component' />
```

</p>

- As URL use `https://github.com/backstage/demo/blob/master/catalog-info.yaml`.

>   This is used by our demo site.

Reviewer: ```suggestion
  This is used by our [demo site](https://demo.backstage.io).
```

<p align='center'>

>   <img src='../assets/getting-started/b-existing-2.png' alt='Register a new component wizard, asking for an URL to the existing component YAML file'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/b-existing-2.png' alt='Register a new component wizard, asking for an URL to the existing component YAML file' />
```

</p>
- Hit `Analyze` and review the changes. Apply them if correct

<p align='center'>

>   <img src='../assets/getting-started/b-existing-3.png' alt='Register a new component wizard, showing the metadata for the component YAML we use in this tutorial'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/b-existing-3.png' alt='Register a new component wizard, showing the metadata for the component YAML we use in this tutorial' />
```

</p>

- You should receive a message that your entities have been added.
- If you go back to `Home`, you should be able to find `demo`. You should be
  able to click it and see the details

Create a new component using a software template

- Go to `create` and choose to create a website with the `React SSR Template`
- Type in a name, let's use `tutorial`
- Select the group `group-a` which will own this new website, and go to the next
  step

<p align='center'>

>   <img src='../assets/getting-started/b-scaffold-1.png' alt='Software template deployment input screen asking for a name, the group owning this and a description'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/b-scaffold-1.png' alt='Software template deployment input screen asking for a name, the group owning this and a description' />
```

</p>

> - For the location, we're going to use the default GitHub location.

Reviewer: What is the default github location? I think suggesting users to use their github account should be straightforward.

Author: Hah! Yes, the default = default.. I'll remove it since they can see what's the default. Don't want to confuse.

- As owner, type your GitHub username
- For the repository name, type `tutorial`. Go to the next step

<p align='center'>

>   <img src='../assets/getting-started/b-scaffold-2.png' alt='Software template deployment input screen asking for the github username and name of the new repo to create'>

Reviewer: ```suggestion
  <img src='../assets/getting-started/b-scaffold-2.png' alt='Software template deployment input screen asking for the github username and name of the new repo to create' />
```

- [Read more about Knex](http://knexjs.org/), which is the library we use for
  the database backend

### Setting up authentication

There's multiple authentication providers available for you to use with
Backstage, feel free to follow
[the instructions for adding authentication](../auth/).

For this tutorial we choose to use GitHub, a free service most of you might be
familiar with. For other options, see
[the auth provider documentation](../auth/github/provider#create-an-oauth-app-on-github).

Go to
[https://github.com/settings/applications/new](https://github.com/settings/applications/new)
to create your OAuth App. The `Homepage URL` should point to Backstage's
frontend, in our tutorial it would be `http://127.0.0.1:3000`. The
`Authorization callback URL` will point to the auth backend, which will most
likely be `http://127.0.0.1:7007/api/auth/github/handler/frame`.

<p align='center'>
  <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page' />
</p>

Take note of the `Client ID` and the `Client Secret`. Open `app-config.yaml`,
and add your `clientId` and `clientSecret` to this file. It should end up
looking like this:

> ```

Reviewer 2: ```suggestion
```yaml
```

auth:
  # see https://backstage.io/docs/auth/ to learn about auth providers
  environment: development
  providers:
    github:
      development:
        clientId: YOUR CLIENT ID
        clientSecret: YOUR CLIENT SECRET
```

Backstage will re-read the configuration. If there's no errors, that's great! We
can continue with the last part of the configuration. The next step is needed to
change the sign-in page, this you actually need to add in the source code.

Open `packages/app/src/App.tsx` and below the last `import` line, add:

> ```

Reviewer 2: ```suggestion
```typescript
```

import { githubAuthApiRef } from '@backstage/core-plugin-api';
import { SignInProviderConfig, SignInPage } from '@backstage/core-components';

const githubProvider: SignInProviderConfig = {
 id: 'github-auth-provider',
 title: 'GitHub',
 message: 'Sign in using GitHub',
 apiRef: githubAuthApiRef,
};
```

Search for `const app = createApp({` in this file, and below `apis,` add:

> ```

Reviewer 2: ```suggestion
```typescript
```

- [Using organizational data from GitHub](../integrations/github/org.md)

### Setting up a GitHub Integration

The GitHub integration supports loading catalog entities from GitHub or GitHub
Enterprise. Entities can be added to static catalog configuration, registered
with the catalog-import plugin, or discovered from a GitHub organization. Users
and Groups can also be loaded from an organization. While using GitHub Apps
might be the best way to set up integrations, for this tutorial you'll use a
Personal Access Token.

Create your Personal Access Token by opening the
[the GitHub token creation page](https://github.com/settings/tokens/new). Use a
name to identify this token and put it in the notes field. Choose a number of
days for expiration. If you have a hard time picking a number, we suggest to go
for 7 days, it's a lucky number.

<p align='center'>
  <img src='../assets/getting-started/gh-oauth.png' alt='Screenshot of the GitHub OAuth creation page' />
</p>

Set the scope to your likings. For this tutorial, selecting "repo" should be
enough.

In the `app-config.yaml`, search for `integrations:` and add your token, like we
did in below example:

> ```

Reviewer 2: ```suggestion
```yaml
```

</p>
- Hit `Analyze` and review the changes. Apply them if correct

<p align='center'>
  <img src='../assets/getting-started/b-existing-3.png' alt='Register a new component wizard, showing the metadata for the component YAML we use in this tutorial' />
</p>

- You should receive a message that your entities have been added.
- If you go back to `Home`, you should be able to find `demo`. You should be
  able to click it and see the details

## Create a new component using a software template

- Go to `create` and choose to create a website with the `React SSR Template`
- Type in a name, let's use `tutorial`
- Select the group `group-a` which will own this new website, and go to the next
  step

<p align='center'>
  <img src='../assets/getting-started/b-scaffold-1.png' alt='Software template deployment input screen asking for a name, the group owning this, and a description' />
</p>

- For the location, we're going to use the default
- As owner, type your GitHub username
- For the repository name, type `tutorial`. Go to the next step

<p align='center'>

>   <img src='../assets/getting-started/b-scaffold-2.png' alt='Software template deployment input screen asking for the github username, and name of the new repo to create' />

Reviewer 2: ```suggestion
  <img src='../assets/getting-started/b-scaffold-2.png' alt='Software template deployment input screen asking for the GitHub username, and name of the new repo to create' />
```

this may fix your markdown spell check error in the pipeline
